File: backend/routers/notifications.py
from fastapi import APIRouter, Depends, Body, HTTPException, Query
from typing import Dict
from auth.clerk import is_clerk_user
from entities.user import User
from services.notification_manager import NotificationManager
from gateways.users import UsersGateway
from entities.message import Message
from gateways.database.mongodb import MongoDBGateway
from ai.assistant.memory import DatabaseMemory
from http import HTTPStatus
from pydantic import BaseModel
from fastapi.responses import Response
from fastapi import Request
from typing import Optional
from constants import (
    SCHEDULED_NOTIFICATION_TIME_DEVIATION_IN_HOURS,
)
from analytics.posthog import posthog
import traceback
from loguru import logger

router = APIRouter()

notification_manager = NotificationManager()
users_gateway = UsersGateway()


class PwaStatusUpdate(BaseModel):
    is_pwa_installed: Optional[bool] = None
    is_pwa_notifications_enabled: Optional[bool] = None
    pwa_subscription_endpoint: Optional[str] = None
    pwa_subscription_key: Optional[str] = None
    pwa_subscription_auth_token: Optional[str] = None


@router.post("/update-pwa-status")
async def update_pwa_status(
    status_update: PwaStatusUpdate = Body(...), user: User = Depends(is_clerk_user)
):
    update_fields = {k: v for k, v in status_update.dict().items() if v is not None}
    updated_user = users_gateway.update_fields(user.id, update_fields)
    return {"message": "PWA status updated successfully", "user": updated_user}


@router.post("/process-scheduled-notification")
async def process_scheduled_notification(request: Request):
    body = await request.json()
    notification_id = body.get("notification_id", None)

    if not notification_id:
        raise HTTPException(status_code=400, detail="Notification ID is required")

    processed_notification = await notification_manager.process_notification(
        notification_id
    )

    if processed_notification:
        user = users_gateway.get_user_by_id(processed_notification.user_id)

        # Send push notification
        try:
            notification_manager.send_push_notification(
                user.id,
                title=f"hey {user.name}",
                body=processed_notification.message.lower(),
                url=f"/add?notification_id={processed_notification.id}",
            )
            logger.info(f"Sent push notification to {user.id}")
        except Exception as e:
            logger.error(f"Failed to send push notification: {e}")
            raise Exception(f"Failed to send push notification: {e}")

        memory = DatabaseMemory(MongoDBGateway("messages"), user.id)
        memory.write(
            Message.new(
                text=processed_notification.message,
                sender_name="Jarvis",
                sender_id="0",
                recipient_name=user.name,
                recipient_id=user.id,
            )
        )

        if processed_notification.type == "engagement":
            posthog.capture(
                distinct_id=user.id,
                event="scheduled-engagement-notification-processed",
                properties={
                    "notification_id": processed_notification.id,
                },
            )
        return {"message": "Notification processed and sent successfully"}
    else:
        return Response(status_code=HTTPStatus.NO_CONTENT.value, content="No notification processed")


@router.post("/mark-notification-opened")
async def mark_notification_opened(
    notification_id: str = Query(...), user: User = Depends(is_clerk_user)
):
    notification = notification_manager.get_notification(notification_id)

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    if notification.user_id != user.id:
        raise HTTPException(
            status_code=403, detail="Not authorized to mark this notification as opened"
        )

    updated_notification = notification_manager.mark_as_opened(notification_id)

    return {
        "message": "Notification marked as opened",
        "notification": updated_notification,
    }


@router.get("/load-notifications")
async def load_notifications(user: User = Depends(is_clerk_user)):
    notifications = [
        n
        for n in notification_manager.get_all_for_user(user.id)
        if n.status != "concluded"
    ]
    for notification in notifications:
        if notification.status == "processed":
            notification_manager.mark_as_opened(notification.id)
    return {"notifications": notifications}


@router.post("/conclude-notification/{notification_id}")
async def conclude_notification(
    notification_id: str, user: User = Depends(is_clerk_user)
):
    notification = notification_manager.get_notification(notification_id)
    if not notification or notification.user_id != user.id:
        raise HTTPException(status_code=404, detail="Notification not found")

    concluded_notification = notification_manager.conclude_notification(notification_id)
    return {
        "message": "Notification concluded",
        "notification_id": concluded_notification.id,
    }


@router.post("/trigger-push-notification")
async def trigger_push_notification(
    request: Request, user: User = Depends(is_clerk_user)
):
    try:
        body = await request.json()

        title = body.get("title", None)
        body = body.get("body", None)
        url = body.get("url", None)
        icon = body.get("icon", None)

        return await notification_manager.send_push_notification(
            user.id, title, body, url, icon
        )
    except Exception as e:
        logger.error(f"Failed to send push notification: {e}")
        logger.error(traceback.format_exc())
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/clear-all-notifications")
async def clear_all_notifications(user: User = Depends(is_clerk_user)):
    notifications = notification_manager.get_all_for_user(user.id)
    
    # Find the latest engagement notification
    latest_engagement = None
    for notification in notifications:
        if notification.type == "engagement" and notification.status != "concluded":
            if not latest_engagement or notification.created_at > latest_engagement.created_at:
                latest_engagement = notification
    
    # Conclude all except latest engagement
    for notification in notifications:
        if notification.status != "concluded":
            if notification != latest_engagement:
                notification_manager.conclude_notification(notification.id)
                
    return {"message": "All notifications cleared except latest engagement"}

File: apps/backend-node/src/routes/notifications.ts
import { Request, Response, Router } from "express";
import { AuthenticatedRequest, requireAuth } from "../middleware/auth";
import { memoryService } from "../services/memoryService";
import { notificationService } from "../services/notificationService";
import { logger } from "../utils/logger";
import { prisma } from "../utils/prisma";
import { MessageRole } from "@tsw/prisma";
import { chatService } from "@/services/chatService";

const router = Router();

// Process scheduled notification
router.post(
  "/process-scheduled-notification",
  async (req: Request, res: Response): Promise<Response | void> => {
    try {
      const { notification_id } = req.body;

      if (!notification_id) {
        return res.status(400).json({ error: "Notification ID is required" });
      }

      logger.info(`Processing scheduled notification ${notification_id}`);

      // Get the notification
      const notification = await prisma.notification.findUnique({
        where: { id: notification_id },
        include: { user: true },
      });

      if (!notification) {
        return res.status(404).json({ error: "Notification not found" });
      }

      // Update notification status to processed and set processedAt
      const processedNotification = await prisma.notification.update({
        where: { id: notification_id },
        data: {
          status: "PROCESSED",
          processedAt: new Date(),
        },
      });

      // Send push notification if user has PWA notifications enabled
      if (
        notification.user.isPwaNotificationsEnabled &&
        notification.user.pwaSubscriptionEndpoint
      ) {
        try {
          await notificationService.sendPushNotification(
            notification.user.id,
            `hey ${notification.user.name || notification.user.username}`,
            processedNotification.message.toLowerCase(),
            `/add?notification_id=${processedNotification.id}`
          );
          logger.info(`Sent push notification to ${notification.user.id}`);
        } catch (pushError) {
          logger.error("Failed to send push notification:", pushError);
          throw pushError;
        }
      }

      const { chat } = await chatService.ensureNotificationsCoachExists(
        notification.userId
      );

      // Store message in conversation memory
      await memoryService.writeMessage({
        content: processedNotification.message,
        chatId: chat.id,
        role: MessageRole.COACH,
      });

      res.json({ message: "Notification processed and sent successfully" });
    } catch (error) {
      logger.error("Error processing scheduled notification:", error);
      res.status(500).json({ error: "Failed to process notification" });
    }
  }
);

// Mark notification as opened (legacy single notification endpoint)
router.post(
  "/mark-notification-opened",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const { notification_id } = req.query;

      if (!notification_id) {
        return res.status(400).json({ error: "notification_id is required" });
      }

      const notification = await prisma.notification.findUnique({
        where: { id: notification_id as string },
      });

      if (!notification) {
        return res.status(404).json({ error: "Notification not found" });
      }

      if (notification.userId !== req.user!.id) {
        return res.status(403).json({
          error: "Not authorized to mark this notification as opened",
        });
      }

      const updatedNotification = await prisma.notification.update({
        where: { id: notification_id as string },
        data: {
          status: "OPENED",
          openedAt: new Date(),
        },
      });

      logger.info(`Marked notification ${notification_id} as opened`);
      res.json({
        message: "Notification marked as opened",
        notification: updatedNotification,
      });
    } catch (error) {
      logger.error("Error marking notification as opened:", error);
      res.status(500).json({ error: "Failed to mark notification as opened" });
    }
  }
);

// Mark multiple notifications as opened (batch endpoint)
router.post(
  "/mark-notifications-opened",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const { notification_ids } = req.body;

      if (!notification_ids || !Array.isArray(notification_ids)) {
        return res
          .status(400)
          .json({ error: "notification_ids array is required" });
      }

      if (notification_ids.length === 0) {
        return res.json({
          message: "No notifications to mark as opened",
          count: 0,
        });
      }

      // Verify all notifications belong to the authenticated user
      const notifications = await prisma.notification.findMany({
        where: {
          id: { in: notification_ids },
        },
        select: { id: true, userId: true },
      });

      // Check if any notifications don't belong to the user
      const unauthorizedNotifications = notifications.filter(
        (n) => n.userId !== req.user!.id
      );

      if (unauthorizedNotifications.length > 0) {
        return res.status(403).json({
          error: "Not authorized to mark some notifications as opened",
        });
      }

      // Only update notifications that actually exist and belong to the user
      const validNotificationIds = notifications.map((n) => n.id);

      if (validNotificationIds.length === 0) {
        return res.status(404).json({ error: "No valid notifications found" });
      }

      // Batch update all valid notifications
      const result = await prisma.notification.updateMany({
        where: {
          id: { in: validNotificationIds },
          userId: req.user!.id,
        },
        data: {
          status: "OPENED",
          openedAt: new Date(),
        },
      });

      logger.info(
        `Marked ${result.count} notifications as opened for user ${req.user!.id}`
      );
      res.json({
        message: "Notifications marked as opened",
        count: result.count,
      });
    } catch (error) {
      logger.error("Error marking notifications as opened:", error);
      res.status(500).json({ error: "Failed to mark notifications as opened" });
    }
  }
);

router.get(
  "/",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const notifications = await prisma.notification.findMany({
        where: {
          userId: req.user!.id,
        },
        orderBy: {
          createdAt: "desc",
        },
      });
      return res.json(notifications);
    } catch (error) {
      logger.error("Error loading notifications:", error);
      res.status(500).json({ error: "Failed to load notifications" });
    }
  }
);

router.post(
  "/conclude/:notificationId",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const { notificationId } = req.params;

      const notification = await prisma.notification.findUnique({
        where: { id: notificationId },
      });

      if (!notification || notification.userId !== req.user!.id) {
        return res.status(404).json({ error: "Notification not found" });
      }

      const concludedNotification = await prisma.notification.update({
        where: { id: notificationId },
        data: {
          status: "CONCLUDED",
          concludedAt: new Date(),
        },
      });

      logger.info(`Concluded notification ${notificationId}`);
      res.json({
        message: "Notification concluded",
        notification_id: concludedNotification.id,
      });
    } catch (error) {
      logger.error("Error concluding notification:", error);
      res.status(500).json({ error: "Failed to conclude notification" });
    }
  }
);

// Trigger push notification (for testing)
router.post(
  "/trigger-push-notification",
  requireAuth,
  async (req: AuthenticatedRequest, res: Response) => {
    try {
      const { title, body, url, icon, isNative } = req.body;

      const result = await notificationService.sendPushNotification(
        req.user!.id,
        title,
        body,
        url,
        icon
      );

      res.json(result);
    } catch (error) {
      logger.error("Error triggering push notification:", error);
      res.status(500).json({ error: "Failed to send push notification" });
    }
  }
);

// Clear all notifications except latest engagement
router.post(
  "/clear-all-notifications",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      // Get all non-concluded notifications
      const notifications = await prisma.notification.findMany({
        where: {
          userId: req.user!.id,
          status: { not: "CONCLUDED" },
        },
        orderBy: { createdAt: "desc" },
      });

      // Find the latest engagement notification
      const engagementNotifications = notifications.filter(
        (n) => n.type === "ENGAGEMENT"
      );
      const latestEngagement =
        engagementNotifications.length > 0 ? engagementNotifications[0] : null;

      // Conclude all notifications except the latest engagement and action-required notifications
      const notificationsToClose = notifications.filter(
        (n) =>
          n.id !== latestEngagement?.id &&
          n.type !== "FRIEND_REQUEST" &&
          n.type !== "PLAN_INVITATION"
      );

      if (notificationsToClose.length > 0) {
        await prisma.notification.updateMany({
          where: {
            id: { in: notificationsToClose.map((n) => n.id) },
          },
          data: {
            status: "CONCLUDED",
            concludedAt: new Date(),
          },
        });
      }

      logger.info(
        `Cleared ${notificationsToClose.length} notifications for user ${req.user!.id}`
      );
      res.json({
        message:
          "All notifications cleared except action-required notifications",
      });
    } catch (error) {
      logger.error("Error clearing notifications:", error);
      res.status(500).json({ error: "Failed to clear notifications" });
    }
  }
);

// Get PWA subscription
router.get(
  "/get-pwa-subscription",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const user = await prisma.user.findUnique({
        where: { id: req.user!.id },
        select: { pwaSubscriptionEndpoint: true },
      });

      res.json({ stored_endpoint: user?.pwaSubscriptionEndpoint || null });
    } catch (error) {
      logger.error("Error getting PWA subscription:", error);
      res.status(500).json({ error: "Failed to get PWA subscription" });
    }
  }
);

// Create and immediately process notification
router.post(
  "/create-and-process-notification",
  requireAuth,
  async (
    req: AuthenticatedRequest,
    res: Response
  ): Promise<Response | void> => {
    try {
      const {
        message,
        type,
        relatedId,
        relatedData,
        promptTag,
        pushNotify = true,
      } = req.body;

      if (!message) {
        return res.status(400).json({ error: "Message is required" });
      }

      const notification =
        await notificationService.createAndProcessNotification(
          {
            userId: req.user!.id,
            message,
            type: type || "INFO",
            relatedId,
            relatedData,
            promptTag,
          },
          pushNotify
        );

      logger.info(
        `Created and processed notification for user ${req.user!.id}`
      );
      res.json({
        message: "Notification created and processed successfully",
        notification,
      });
    } catch (error) {
      logger.error("Error creating and processing notification:", error);
      res
        .status(500)
        .json({ error: "Failed to create and process notification" });
    }
  }
);

// Get latest notification sent to user (optionally filtered by type)
router.get(
  "/latest-sent",
  requireAuth,
  async (req: AuthenticatedRequest, res: Response) => {
    try {
      const { type } = req.query;

      const notification =
        await notificationService.getLatestNotificationSentToUser(
          req.user!.id,
          type as any
        );

      res.json({ notification });
    } catch (error) {
      logger.error("Error getting latest sent notification:", error);
      res.status(500).json({ error: "Failed to get latest sent notification" });
    }
  }
);

export const notificationsRouter: Router = router;
export default notificationsRouter;
